scripts/common.scala

```diff
@@ -145,53 +147,55 @@ object Progress:
-  val first   = LocalTime.now()
-
-  def elapsed(from: Temporal, to: Temporal = LocalTime.now()): String = {
-    Duration.between(from, to).toSeconds match
-      case s if s < 60 => s"$s seconds"
-      case s           => s"${s / 60} minutes"
-  }
+  val counter   = new AtomicInteger(0)
+  val failed    = collection.concurrent.TrieMap.empty[String, String]
+  val succeeded = collection.concurrent.TrieMap.empty[String, String]
+  val first     = LocalTime.now()
+
+  def elapsed(from: Temporal, to: Temporal = LocalTime.now()): String =
+    Duration.fromNanos(java.time.Duration.between(from, to).toNanos) match
+      case d if d.toHours > 0   => f"${d.toHours} h ${d.toMinutes % 60} m ${d.toSeconds % 60} s"
+      case d if d.toMinutes > 0 => f"${d.toMinutes} m ${d.toSeconds % 60} s"
+      case d if d.toSeconds > 0 => f"${d.toSeconds} s ${d.toMillis % 1000} ms"
+      case d if d.toMillis > 0  => f"${d.toMillis} ms ${d.toMicros % 1000} µs"
+      case d                    => f"${d.toMicros} µs"
+
+  def end(lbl: String, start: Temporal, end: Temporal) =
+    if !failed.contains(lbl)
+    then succeeded.put(lbl, s"\r$lbl: DONE [${elapsed(start, end)}]")
+  def failure(lbl: String, start: Temporal, end: Temporal, error: String) = 
+    failed.put(lbl, s"\r$lbl: ERROR [${elapsed(start, end)}]: $error")
+
+  def report(lbl: String, amount: Int, time: Temporal): Unit =
+    val current = counter.addAndGet(amount)
+    val percentage = (current.toDouble / total * 100).toInt
+    Console.out.synchronized {
+      println(s"\r$lbl: $percentage% [$current/$total] [${time}]")
+    }
 
   // noinspection ScalaUnusedSymbol
-  val progress = Progress(
-    title,
-    report = (lbl, amount, time) =>
-      val current    = counter.addAndGet(amount)
-      val percentage = (current.toDouble / total * 100).toInt
-      Console.out.synchronized {
-        println(s"\r$lbl: $percentage% [$current/$total] [${time}]")
-      }
-    ,
-    summary = (lbl, start, end) =>
-      if failed.contains(lbl) then
-        Console.out.synchronized {
-          println(s"\r$lbl: ERROR [${elapsed(start, end)}]")
-        }
-      else
-        Console.out.synchronized {
-          println(s"\r$lbl: DONE [${elapsed(start, end)}]")
-        }
-    ,
-    failure = (lbl, error) =>
-      failed.put(lbl, error)
-  )
+  val progress = Progress(title, report = report, ending = end, failure = failure)
 
   println(title)
 
   try f(using progress)
   finally
-    println(s"Total time: ${elapsed(first)}")
+    println()
+    println(s"Successes [${succeeded.size}/$total]:")
+    succeeded.toSeq.sortBy(_._1).foreach((name, msg) => println(s"  - $name: $msg"))
+    println()
     if failed.nonEmpty then
-      println(s"Failures [${failed.size}]:")
-      failed.foreach((name, error) => println(s"  - $name: $error"))
+      println(s"Failures [${failed.size}/$total]:")
+      failed.toSeq.sortBy(_._1).foreach((name, error) => println(s"  - $name: $error"))
```

Comment:
why do we need to sort here? also fuck toSeq, use toVector, lists suck

Reply:
I'm sorting to get consistent display in alphabetic order